dot/core/service.go

```diff
@@ -191,7 +191,15 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 		block.Header.Hash(), state.MustRoot())
 
 	rt, err := s.blockState.GetRuntime(&block.Header.ParentHash)
-	if err != nil {
+	if errors.Is(err, blocktree.ErrFailedToGetRuntime) {
+		rt, err = s.storageState.GetRuntime(block.Header.ParentHash)
+		if err != nil {
+			return fmt.Errorf("getting runtime from database: %w", err)
+		}
+		// ensure the runtime stops and releases resources since it was
+		// instantiated from disk just for this function call.
```

Comment:
Why is it the responsibility of the caller of `BlockState.GetRuntime` to get the runtime from the `StorageState` if it doesn't exist? Why doesn't it exist?

Reply:
It may not exist in BlockState because it's not the current runtime.  This was updated based on request from this comment: https://github.com/ChainSafe/gossamer/pull/2781#pullrequestreview-1099168287

Reply:
Why is the responsibility of the caller to manage this? I'm still not sure why the runtime wouldn't exist.  Where should the runtimes exist? The block tree or the storage state? It shouldn't be on the onus of the caller to have to manage this.  Why can't we just retrieve the runtime from whatever is supposed to contain it, and leave the caching in memory to that package.